### ISSUE TITLE
Durable Entity State Seems to be Getting Lost Possibly Cause Competing Hosts Saving State During Shutdown / DrainMode enabled

### ISSUE THREAD

**Issue by drdamour** [Azure/azure-functions-durable-extension#2437]
### Description
We have a durable entity that thottles concurrent delivery of incoming messages from an event hub to a target.  It does this with an EH binding putting Signaling the entity with an Enque, and then an eternal orchestration Once every couple of mins Calls() the process method n the entity which drains the queue.

Once in a while...messages seems to disappear from the queue.  EG we have 980 messages for receiving the event message from hub, 980 messages about "EntitySignal:Queue" but when our process method kicks off it only sees 808 messages in the queue, and we only see 808 processing messages. 

~Nothing is standing out in the logs~, and we've been unable to recreate this problem anywhere but prod :(

update: found some unexpected behaviour in log, see first comment

looking for any advice/things to check/way to verify somethings going wrong with the entity state (probably because we are doing something wrong)

```
public class ThrottleSender
{
    //it's a singleton so we will always use it a single Id
    public const string InstanceId = "7399a8cd-878d-48a2-a37c-684210bcb902";
    public static readonly EntityId SingletonEntityId = new EntityId(
        nameof(ThrottleSender),
        InstanceId
    );


    [FunctionName(nameof(ThrottleSender))]
    public static Task Run(
        [EntityTrigger] IDurableEntityContext ctx
    ) => ctx.DispatchAsync<ThrottleSender>();


    [JsonProperty]
    private Queue<EnqueuedCommand> queuedCommands = new Queue<EnqueuedCommand>();

    [JsonProperty]
    private List<FailedCommand> failedCommands = new List<FailedCommand>();

    [JsonProperty]
    private List<SuccessfulCommand> recentlySuccessfulCommands = new List<SuccessfulCommand>();


    [JsonIgnore]
    private readonly ServiceClient serviceClient;

    [JsonIgnore]
    private readonly ILogger<ThrottleSender> log;

    [JsonIgnore]
    private readonly TelemetryClientWrapper telemetryClient;

    public ThrottleSender(
        ServiceClient serviceClient,
        ILogger<ThrottleSender> log,
        TelemetryClientWrapper telemetryClient
    )
    {
        this.serviceClient = serviceClient;
        this.log = log;
        this.telemetryClient = telemetryClient;
    }


    public Task<ThrottleSender> Queue(
        DoStuffCommand command
    )
    {
        queuedCommands.Enqueue(
            new EnqueuedCommand(
                Guid.NewGuid(),
                DateTimeOffset.UtcNow,
                command
            )   
        );
        return Task.FromResult(this);
    }

    public async Task<ThrottleSender> ProcessQueuedCommands(
    )
    {

        // reset as we want to track this per run
        requestCount = 0;

        var startTime = DateTimeOffset.UtcNow;
       //stay under consumption plan 10 mins
        var endTime = startTime + TimeSpan.FromMinutes(5);

        var enqueuedCommandsBeforeMetric = new MetricTelemetry()
        {
            Name = Troubleshooting.TelemetryMetricNames.QUEUED_COMMANDS,
            Sum = queuedCommands.Count(),
            Timestamp = startTime,
        };

        var failedCommandsBeforeMetric = new MetricTelemetry()
        {
            Name = Troubleshooting.TelemetryMetricNames.FAILED_COMMMANDS_RETAINED,
            Sum = failedCommands.Count(),
            Timestamp = startTime,
        };

        var successCommandsBeforeMetric = new MetricTelemetry()
        {
            Name = Troubleshooting.TelemetryMetricNames.SUCCESSFUL_COMMANDS_RETAINED,
            Sum = recentlySuccessfulCommands.Count(),
            Timestamp = startTime,
        };

        telemetryClient.TrackMetric(enqueuedCommandsBeforeMetric);
        telemetryClient.TrackMetric(failedCommandsBeforeMetric);
        telemetryClient.TrackMetric(successCommandsBeforeMetric);

        //TODO: is this needed?
        telemetryClient.Flush();

        var sequentialFailures = 0;

        var inFlightTasks = new List<Task<DoStuffCommandResult>>();
        var inFlightCommands = new List<EnqueuedCommand>();
        var maxConcurrency = 5;

        // run loop while we have time AND (commandsToAdd OR processesToFinish)
        while (DateTimeOffset.UtcNow < endTime && (queuedCommands.Count > 0 || inFlightCommands.Count > 0)) 
        {
           //fill up with concurrent tasks
            while (DateTimeOffset.UtcNow < endTime && queuedCommands.Count > 0 && inFlightTasks.Count < maxConcurrency)
            {
                var c = queuedCommands.Dequeue();

                inFlightTasks.Add(serviceClient.Execute(c));
                inFlightCommands.Add(c);
            }

            var finishedTask = await Task.WhenAny(inFlightTasks);

            inFlightTasks.Remove(finishedTask);

            var result = await finishedTask;

            // remove the command whether or not it succeeded
            inFlightCommands.Remove(result.Command);

            if(result.Exception != null)
            {
                failedCommands.Add(
                    new FailedCommand(
                        result.Command.Id,
                        result.Command.Enqueued,
                        result.Command.Command,
                        DateTimeOffset.UtcNow,
                        result.Exception
                    )
                );

                log.LogError(
                    Troubleshooting.EventIDs.REQUEST_FAILURE,
                    result.Exception,
                    "executing command failed"
                );

                sequentialFailures++;

                if (sequentialFailures > maxSequentialFailures)
                {
                    // TODO: do we need to reset this? don't want the entity to sit around in dirty state
                    sequentialFailures = 0;

                    throw new Exception(
                        "ProcessQueuedCommands sequential failure limit reached"
                    );
                }

                continue;

            }

            // if we got here, task succeeded
            sequentialFailures = 0;

            log.LogInformation(
                    "processed command"
                );

            recentlySuccessfulCommands
            .Add(
                new SuccessfulCommand(
                    result.Command.Id,
                    result.Command.Enqueued,
                    result.Command.Command,
                    DateTimeOffset.UtcNow
                )
            );

        }

        //HACK: throw the rest back in the start of queue, we can redeliver these
        foreach (var c in inFlightCommands.Reverse())
        {
            queuedCommands = new Queue<EnqueuedCommand>(queuedCommands.Prepend(c));
        }

        // Hold onto recent success for 3 days
        for (var i = 0; i < recentlySuccessfulCommands.Count; i++)
        {
            if (recentlySuccessfulCommands[i].Succeeded < startTime.AddDays(-3))
            {
                recentlySuccessfulCommands.RemoveAt(i);
                break;
            }
        }

        // endtime doesn't always == time runout
        endTime = DateTimeOffset.UtcNow;

        var enqueuedCommandsAfterMetric = new MetricTelemetry()
        {
            Name = Troubleshooting.TelemetryMetricNames.QUEUED_COMMANDS,
            Sum = queuedCommands.Count(),
            Timestamp = endTime,
        };

        var failedCommandsAfterMetric = new MetricTelemetry()
        {
            Name = Troubleshooting.TelemetryMetricNames.FAILED_COMMMANDS_RETAINED,
            Sum = failedCommands.Count(),
            Timestamp = endTime,
        };

        var successCommandsAfterMetric = new MetricTelemetry()
        {
            Name = Troubleshooting.TelemetryMetricNames.SUCCESSFUL_COMMANDS_RETAINED,
            Sum = recentlySuccessfulCommands.Count(),
            Timestamp = endTime,
        };

        var requestCountMetrics = new MetricTelemetry()
        {
            Name = Troubleshooting.TelemetryMetricNames.REQUEST_SENT,
            Sum = requestCount,
            Timestamp = endTime,
        };

        telemetryClient.TrackMetric(enqueuedCommandsAfterMetric);
        telemetryClient.TrackMetric(failedCommandsAfterMetric);
        telemetryClient.TrackMetric(successCommandsAfterMetric);
        telemetryClient.TrackMetric(requestCountMetrics);

        return await Task.FromResult(this);
    }

}

public record EnqueuedCommand(
    Guid Id,
    DateTimeOffset Enqueued,
    DoStuffCommand Command 
);

```

the thing filling the queue is DEAD simple

```
[FunctionName("QueueStuff")]
    public async Task QueueStuff(
        [EventHubTrigger("stuff", Connection = "con", ConsumerGroup = "$notdefault")] EventData message,
        [DurableClient] IDurableEntityClient durableEntityClient
    )
    {
        //bunch of value checks omitted

              await durableEntityClient.SignalEntityAsync(
                  ThottleSender.SingletonEntityId,
                  nameof(ThottleSender.Queue),
                  message.AsCommand()
              );
    }
```

and the eternal orchestration is

```
    [FunctionName(nameof(ProcessQueuedCommands))]
    public async Task ProcessQueuedCommands(
        [OrchestrationTrigger] IDurableOrchestrationContext context
    )
    {

        var replaySafelogger = context.CreateReplaySafeLogger(log);
        var loop = 0;
        
        while (loop++ < 100)
        {

            await context.CallEntityAsync(
                ThrottleSender.SingletonEntityId,
                nameof(ThrottleSender.ProcessQueuedCommands)
            );

            var completeTime = context.CurrentUtcDateTime;


            await context.CreateTimer(completeTime.AddSeconds(10), CancellationToken.None);
            
        }

        //after 100 loops we want to clear history, but keep executing see best practice at https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-eternal-orchestrations?tabs=csharp
        context.ContinueAsNew();
    }
```

### Expected behavior
Expect if we see 980 SignalEntity:Queue messages we'd see 980 commands in the queue when process start is called

### Actual behavior
only saw 808 commmands in the queue

### Relevant source code snippets
see above

### Known workarounds
none

### App Details

- **Durable Functions extension version (e.g. v1.8.3)**: 2.7.2
- **Azure Functions runtime version (1.0 or 2.0)**: v4 .net 6 in process
- **Programming language used**: c#

### Screenshots
> If applicable, add screenshots to help explain your problem.

### If deployed to Azure
> We have access to a lot of telemetry that can help with investigations. Please provide as much of the following information as you can to help us investigate!

- **Timeframe issue observed**: 8am -9am MT
- **Function App name**:  fgp01fap-amc-ont-ext-prod
- **Function name(s)**: holdreleasecommandprocessor
- **Azure region**: West US 2
- **Orchestration instance ID(s)**:  @holdreleasecommandprocessor@7399a8cd-878d-48a2-a37c-684210bcb902 is entity id
- **Azure storage account name**: fgp01stramcontextprod

> If you don't want to share your Function App or storage account name GitHub, please at least share the orchestration instance ID. Otherwise it's extremely difficult to look up information.

**Comment by drdamour** [comment]
ok spent a few more hours trolling the logs...there was a minute 8:49 were exactly 172 queue method were invoke...so that seemed like a bad coincidence.   looking around those logs...this entity was processing things simultaniously on two hosts...as i understood it that is NOT supposed to happen.

during the execution of ProcessQueuedCommands these logs came thru
![image](https://user-images.githubusercontent.com/1514496/230285496-fc0af8ec-adda-45d5-beb8-aa9eccfe7870.png)

so that host was being shut down

shortly after that the other host started the entity

![image](https://user-images.githubusercontent.com/1514496/230287958-e84a4b23-2e4f-41dc-b2c0-cbead790f272.png)

but during that the first host (that was shutting down) was happy to keep running it's activity...

![image](https://user-images.githubusercontent.com/1514496/230288353-dd53f972-57dd-4531-a7be-da2ee6b45ac5.png)


until it finally exited, and the host quickly shut down

![image](https://user-images.githubusercontent.com/1514496/230288550-ca0f17e6-84db-4b3e-83c2-3aad754fd4a5.png)

and presumably that shutting down host then serialized it's state (which was now an empty queue) to storage...and all the state from the other host was wiped out.

so...
(1) is this a bug in the runtime?  
(2) should two hosts ever be working with the same entity instance id simultaniously?
(3) are we doing something wrong in our code?    like are we supposed to take a cancellation token and have our activity early exit because of the drain event?  i don't recall ever reading any docs around that...but it kinda makes sense....
(4) any other advice?

**Comment by drdamour** [comment]
took a look at #1417 which seems very related, but instead of it being a problem at start up this seems to be a problem at shutdown.

found mention more of graceful shutdown #1112 , but not seeing any docs on what that really means. i see there's a flag, but not sure that helps or hurts for this bug.  seems like it's about not failing the running activity...but my problem looks to be that it did NOT fail.

**Comment by sebastianburckhardt** [comment]
Supposedly, the state of an entity is protected from split-brain problems by the e-tag. The mechanism is supposed to work like this: If there are temporarily two hosts processing entity messages at the same time, one of them is forced to fail when writing back the state to storage, and at that point will not mark the message as consumed, so it can be redelivered to the true host.

That said, it is of course possible something went wrong here. We will need to investigate.

**Comment by drdamour** [comment]
we had another instance of this, we updated prior to the latest df extension 2.9.3 as our MS support rep asked us to, and we instrumented a lot more logging and can see messages from both entities where the counts are "split brain"

we also enabled graceful shutdown in config and hoped this would mean the cancellation token would be triggered, but it appears that it was NOT so our loop did not early exit so our assumptions regarding #1112 seems wrong...

**Comment by sebastianburckhardt** [comment]
FYI, I have looked that the split-brain instance you mentioned above (4/5/2023), and based on the traces it looked like it was handled correctly (the host on 10.30.1.167 detects an e-tag violation when writing the state back, which correctly cancels the write, and abandons the message in question). So this particular split-brain occurrence does not explain why some messages were not processed.

My next step is to try to see if I can use a query to find any unprocessed messages.

**Comment by sebastianburckhardt** [comment]
> can see messages from both entities where the counts are "split brain"

It is unfortunately not something we can prevent, since the partition manager cannot always detect failures accurately. The goal is therefore not to prevent split brain completly, but to correctly recover from it. That is, only one of the two entity copies will commit its state to storage and the other is going to get "rolled back". I realize that this can be confusing when looking at the logs since you may see a count that "goes backwards". However, the rolled-back entity also puts the messages back into the queue, so they can get reprocessed, so ultimately all messages should always be processed. It makes this tricky to troubleshoot unfortunately.

I have tried using queries to find messages that were not processed. I wrote a query that examines all the messages in the traces between 2023-04-05 and 2023-04-18, but couldn't find any indication that a message was processed less than once or more than once.

**Comment by drdamour** [comment]
Yeah we weren't looking to avoid the logs showing different values. We only brought that up as evidence that two entities are running side by side both with their own copies of our queue.



to be clear its not the df messages that are lost.  We see all those be processed.  But the state being written by the one that accumulates the messages in our queue is being lost in favor of the one that eats the messages. So again its our messages in our queue being lost.

you mentioned you saw the etag error, which function had the error?  The feeders or the eater?  Since the feeders are batched if they are the one that failed to write then that would lead to the same situation, we see the feeders execute but the final state is not persisted because the zombie hosts writes to storage.  So no overwrite happens but we still miss out.

**Comment by sebastianburckhardt** [comment]
> But the state being written by the one that accumulates the messages in our queue is being lost in favor of the one that eats the messages. So again its our messages in our queue being lost.

I am not sure that I fully understand the scenario you are describing, but let's see if I can figure it out by talking it through. I am going to describe how the system is designed to handle this type of scenario (which could be different from what is actually happening if there is a bug, of course).

Let's say we have a split brain for the singleton entity where 
 - host A executes `Queue` operations, which adds elements to the `queuedCommands` field. 
 - host B executes `ProcessQueuedCommands` operation, which removes elements from `queuedCommands` and adds them to `recentlySuccessfulCommands` or `failedCommands`.
 Now let's say one of these hosts loses the storage race. What will happen?

1. *If host A loses*: the entity state is rolled back to what it was before the 'Queue' operation, i.e. the elements are no longer in `queuedCommands`. The entity signal will be redelivered at some point, at which point `Queue` executes again and adds those messages to `queuedCommands` again.  This seems pretty harmless.
2. *If host B loses*: the entity state is rolled back to what it was before the 'ProcessQueuedCommands' operation. In particular, the commands that were executed are back in the `queuedCommands` queue, and no longer in the `recentlySuccessfulCommands` or `failedCommands` queues. At some point, the 'ProcessQueuedCommands' signal is redelivered (or the eternal orchestration just sends another one anyway), and 'ProcessQueuedCommands' runs again, and will *process the same commands again*, calling serviceClient again for the same commands, but possibly with a different outcome than before (what previously succeeded may fail this time, or vice versa).

Clearly, the scenario 2 is nastier - because of the *external effects*: the calls to the telemetryClient and serviceClient can't be rolled back when the entity state is rolled back, so the effect is that you are seeing duplicate external calls for the same command. 

At this point I am assuming that scenario 2 is what you are observing. Is that right? Or are you saying that the command processing is not just duplicated, but totally lost?

Just as an aside: duplication of external effects is unfortunately always an issue of workflow systems. The usual recommendation is to ensure that "external effects are idempotent", whatever that means. Well, in this case, it would mean that serviceClient would have to be able to deal correctly with duplicate calls for a command in some way. This is a fundamental problem of workflow systems, it is not a consequence of split brain specifically. The same thing can happen without split brain if host B crashes for whatever reason before writing the entity state to storage. It is also not specific to entities, the same thing can happen with activities.

**Comment by drdamour** [comment]
understand the idempotency, and we are within the bounds of that. delivering/eating the queue repeatedly is fine for our scenario IE redelivery is fine.

the results of neither 1 or 2 is happening to us.   what we are experiencing is closest to scenario 1, because the queue only has the results after the eater (all the queued commands are gone) so it seems like host A lost.  however we never see the queue operation come through again...they seem to be effectively lost.  maybe that's a bug?

**Comment by collins-benj** [comment]
for reference, on 4/16/2023, between 3:35 PM and 3:39 PM UTC, 363 messages were enqueued & ran the `Queue` function noted above, and all were no lost due to this behavior. None of these had a trace to reflect a second run of `Queue`.

**Comment by collins-benj** [comment]
> I have looked that the split-brain instance you mentioned above (4/5/2023), and based on the traces it looked like it was handled correctly (the host on 10.30.1.167 detects an e-tag violation when writing the state back, which correctly cancels the write, and abandons the message in question)

I've looked through our App Insights logs and I'm not seeing anything relevant to this. What can we do in order to see these logs?

**Comment by cgillum** [comment]
> I've looked through our App Insights logs and I'm not seeing anything relevant to this. What can we do in order to see these logs?

We have instructions for how to enable the lower-level Durable Task Framework logging here: https://learn.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-diagnostics?tabs=csharp-inproc#durable-task-framework-logging.

**Comment by sebastianburckhardt** [comment]
> for reference, on 4/16/2023, between 3:35 PM and 3:39 PM UTC, 363 messages were enqueued & ran the Queue function noted above, and all *WERE lost due to this behavior. None of these had a trace to reflect a second run of Queue.

Thanks for the details. I will try to see if I can find some clues for this specific window.

**Comment by sebastianburckhardt** [comment]
I can confirm that scenario 2 is taking place in this window. I can also see that the inputs are large, which means the contents are stored in blobs. My current suspicion is that the e-tag protection may not have taken into account this "large properties spill into blobs" mechanism which could explain why the loser's state ends up winning (because it overwrites the blob) even though its table storage access loses (because the e-tag does not match).

**Comment by sebastianburckhardt** [comment]
I have confirmed that this is the likely cause for the error and filed an issue in the corresponding repository.

**Comment by drdamour** [comment]
@sebastianburckhardt  thats a really good find and very promising.

i dont want to lose sight of the sub part of this issue, that there doesnt seem to be a way for a durable entity to participate in graceful shutdown.  Is that by design?  I suspect i should open another issue…

**Comment by sebastianburckhardt** [comment]
> i dont want to lose sight of the sub part of this issue, that there doesnt seem to be a way for a durable entity to participate in graceful shutdown. Is that by design? I suspect i should open another issue…

Yes, I would suggest to open another issue for questions or feature requests regarding graceful shutdown in entities.

There isn't currently any cancellation token you could use as an indication that the partition is shutting down - the cancellation token that is available via the Azure Functions binding fires only when the host is shutting down.

**Comment by collins-benj** [comment]
We have installed a preview version of this lib (`1.13.7-preview`) given to us through our support ticket to hopefully fix our issue https://github.com/Azure/azure-functions-durable-extension/issues/2437. We have seen the following log:
> @holdreleasecommandprocessor@7399a8cd-878d-48a2-a37c-684210bcb902: Split brain detected: Another worker already updated the history for this instance - the 3 history event result(s) will be discarded

immediately after this, we see this error: 
> DurableTask.AzureStorage.Storage.DurableTaskStorageException: Element 3 in the batch returned an unexpected response code.

At this point, the orchestration that calls our entity halts in the "running" state, and never continues as new as we would expect. To unstick we have had to purge orchestration history and start new. This has happened on 3 occasions over the past week.
